infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationSourceManager.java

```diff
@@ -247,16 +221,61 @@ public LogReplicationEntry receive(LogReplicationEntry message) {
         // Process ACKs from Application, for both, log entry and snapshot sync.
         if(message.getMetadata().getMessageMetadataType() == MessageType.LOG_ENTRY_REPLICATED) {
             log.debug("Log entry sync ACK received on timestamp {}", message.getMetadata().getTimestamp());
+            lastAckedTimestamp.set(message.getMetadata().getTimestamp());
             logReplicationFSM.input(new LogReplicationEvent(LogReplicationEventType.LOG_ENTRY_SYNC_REPLICATED,
                 new LogReplicationEventMetadata(message.getMetadata().getSyncRequestId(), message.getMetadata().getTimestamp())));
         } else if (message.getMetadata().getMessageMetadataType() == MessageType.SNAPSHOT_REPLICATED) {
             log.debug("Snapshot sync ACK received on base timestamp {}", message.getMetadata().getSnapshotTimestamp());
+            lastAckedTimestamp.set(message.getMetadata().getTimestamp());
             logReplicationFSM.input(new LogReplicationEvent(LogReplicationEventType.SNAPSHOT_SYNC_COMPLETE,
                     new LogReplicationEventMetadata(message.getMetadata().getSyncRequestId(), message.getMetadata().getTimestamp())));
         } else {
             log.debug("Received data message of type {} not an ACK", message.getMetadata().getMessageMetadataType());
         }
-
         return null;
     }
+
+    /**
+     * For the given replication runtime, query max stream tail for all streams to be replicated.
+     *
+     * @return max tail of all streams to be replicated for the given runtime
+     */
+    private long getMaxReplicatedStreamsTail() {
+        Map<UUID, Long> tailMap = runtime.getAddressSpaceView().getAllTails().getStreamTails();
+        long maxTail = Address.NON_ADDRESS;
+        for (String s : config.getStreamsToReplicate()) {
+            UUID streamUuid = CorfuRuntime.getStreamID(s);
+            if (tailMap.get(streamUuid) != null) {
+                long streamTail = tailMap.get(streamUuid);
+                maxTail = Math.max(maxTail, streamTail);
+            }
+        }
+        return maxTail;
+    }
+
+    /**
+     * Given a timestamp, calculate how many entries remain to be sent for all replicated streams.
+     *
+     * @param ackedTimestamp
+     * Note: This method of calculating the remaining entries to send is inaccurate because we consider the tail
+     * of all streams(replicated and otherwise) in Corfu and compare it with the current entry sent.  It is possible
+     * that the tail is far ahead but replicated streams have few entries to send.
+     */
+    private int calculateRemainingEntriesToSend(long ackedTimestamp) {
+        long timestamp = getMaxReplicatedStreamsTail();
+        long remainingEntriesToSend = timestamp - ackedTimestamp;
+        int percentDone = (int) (ackedTimestamp/timestamp * 100);
```

Comment:
Looks like remainingEntriesToSend is not used?

Reply:
right.  I intended to use it somehow to calculate the remaining entries in a more accurate way, but it is not needed now.  Will remove.